clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -187,7 +187,7 @@ private void startSequencesAtBeginning(TopicPartition topicPartition, ProducerId
         private static final Comparator<ProducerBatch> PRODUCER_BATCH_COMPARATOR = (b1, b2) -> {
             if (b1.baseSequence() < b2.baseSequence()) return -1;
             else if (b1.baseSequence() > b2.baseSequence()) return 1;
-            else return b1.equals(b2) ? 0 : 1;
+            else return b1.equals(b2) ? 0 : Integer.compare(b1.hashCode(), b2.hashCode());
```

Comment:
Do you think we should override the `equals` and `hashcode` method in  `ProducerBatch` class? Otherwise, the hashcode result could have possibility to be identical and makes the original issue happened again.

Reply:
Hi @showuon, thanks for you comment. I found some trouble when adding `equals` method in `ProducerBatch`. The only two final fields I can use is `createdMs` and `topicPartition`, I don't think it make sense if I use these two fields since they usually have duplicates. Do you have some suggestions? Thanks

Reply:
@artemlivshits done, thanks for your suggestion

Reply:
Make sense. Thanks for the update.

Reply:
This is incorrect @artemlivshits, `Object.hashCode` is not based on object addresses in Java. The hashCode is 32 bits and object addresses can be 64-bits. Also, object addresses resulted in more collisions when used with hashmaps than the default implementation.

I have seen collisions in practice that resulted in a huge amount of work to debug and diagnose. If you're interested, this was the fix https://github.com/squeryl/squeryl/commit/6e43673b273c270a39ba887ec9a07d921b566562

Reply:
In this case, even if hashCode always returned 0, it's as good as the original implementation, where any batch with the same baseSequence would be treated as equal and the consequences are clear and deterministic.  Returning 1 when it's not equal may lead to issues that are super hard to debug -- sometimes an element could be found, and sometimes could not (I'd say only race conditions would beat binary search in unsorted container :-)).
In general -- agree that hashCode is not fully specified, so if we really want to make sure that different batches have truly different identities and keep them ordered, we'd need to come up with a stable way to order them, e.g. stamp each batch with a unique integer.  (In C++ I would've just used the address, I couldn't find a built-in way to order 2 objects in Java).
From the release management perspective, here is my regression ranking:
1. Comparing just baseSequence was there for 3 years, so probably not super severe.
2. This fix partially (fairly reliably in practice) fixes the original problem, no regression.
3. Returning 1 when not equal (#11991) introduces a hard to debug problem.

If I had to pick between the three, 2 is a clear winner -- makes the original behavior better, without introducing regressions.

Reply:
But why do we have to limit ourselves to these 3 options? We can do the equals check and then do the hashCode thing if unequal.

Reply:
If equals is true, then hash codes must be equal too, so comparing hashCode for equal objects would return 0 in that case.  If equals is false, then we do hashCode comparison.  If we fold these 2 cases, then if we just do hashCode comparison, the result is the same as comparing equals first.

Reply:
That's true, we can't rely on hashcode if equals is false either. I'll take a look at the code a bit later, but it should be possible to avoid the latent bug here.

Reply:
From the release perspective, it improves over the previous release, so for now we could take this as is and file a bug to work on a more precise solution in a future release.  We could either add a unique long id (for every new object just increment a static atomic and get a new value) or instead of a producer batch use a HashSet of batches.

Reply:
It turns out that there are a couple of simple solutions, see #12096.